### DISCUSSION TITLE
network/linux-bridge

### DISCUSSION THREAD

**Discussion by giscus** [isno/theByteBook#211] [General]
# network/linux-bridge

构建大规模高可用的分布式系统

https://www.thebyte.com.cn/network/linux-bridge.html

**Comment by Plus-Jia** [comment]
“宿主机根据路由规则再送至 VETP 设备或者 TUN 设备封装成一个新的数据包”，这里VETP应该打错了是VETH吧。
再次感谢作者，读到这里收货颇丰

**Reply by isno** [reply]
这儿单词打错，是 VTEP（VXLAN Tunnel End Points），VXLAN 通信里面的设备。 这块写得不清楚，我再重新改改。

**Comment by XinyuHu123456** [comment]
写得太好了

**Comment by Plus-Jia** [comment]
This is an automatic reply, confirming that your e-mail was received.Thank you ! !Plus_Jia

**Comment by wudilun123** [comment]
# 将 veth 另一端连接到 br0
$ ip link set dev veth0 master br0
这里应为$ ip link set dev veth1 master br0吧

**Reply by isno** [reply]
谢谢，已修改

**Comment by wudilun123** [comment]
网络命名空间名ns1/ns2和net1/net2混用了，3.5.3节也有这个问题

**Reply by isno** [reply]
谢谢，看来我还得再抽精力仔细校对一遍

**Comment by xiongyunhua-star** [comment]
需要添加两个语句，将veth1与veth3(桥接到br0的接口) 设置为up,  ip link set veth1 up ; ip link set veth3 up; 否则的话 br0的状态处于DOWN(因为接口都是Down)

**Comment by Plus-Jia** [comment]
This is an automatic reply, confirming that your e-mail was received.Thank you ! !Plus_Jia

**Comment by bingofather** [comment]
测试下来ping不通

**Reply by bingofather** [reply]
配置如下：
$ ip netns add ns1
$ ip netns add ns2
$ ip link add veth0 type veth peer name veth1
$ ip link add veth2 type veth peer name veth3
ip link add name br0 type bridge
ip link set br0 up
$ ip link set veth0 netns ns1
$ ip link set veth2 netns ns2
$ ip link set dev veth1 master br0
$ ip link set dev veth3 master br0
$ ip netns exec ns1 ip link set veth0 up
$ ip netns exec ns1 ip addr add 172.16.0.1/24 dev veth0
$ ip netns exec ns2 ip link set veth2 up
$ ip netns exec ns2 ip addr add 172.16.0.2/24 dev veth2
ip link set veth1 up
ip link set veth3 up
状态：
       valid_lft forever preferred_lft forever
6: veth1@if7: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br0 state UP group default qlen 1000
    link/ether ee:81:66:b0:e1:54 brd ff:ff:ff:ff:ff:ff link-netns ns1
    inet 172.16.0.3/24 scope global veth1
       valid_lft forever preferred_lft forever
    inet6 fe80::ec81:66ff:feb0:e154/64 scope link 
       valid_lft forever preferred_lft forever
8: veth3@if9: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br0 state UP group default qlen 1000
    link/ether 46:6e:44:c1:a9:65 brd ff:ff:ff:ff:ff:ff link-netns ns2
    inet 172.16.0.4/24 scope global veth3
       valid_lft forever preferred_lft forever
    inet6 fe80::446e:44ff:fec1:a965/64 scope link 
       valid_lft forever preferred_lft forever
10: br0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 1e:a1:61:e4:b7:c2 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::1ca1:61ff:fee4:b7c2/64 scope link 
       valid_lft forever preferred_lft forever
以上在ns1内ping 172.16.0.2不通

**Reply by zjyyzhengli** [reply]
我在本地尝试了可以成功，对比了我的操作，与你的差异是：
下面这两条命令，我是在 `ip link set br0 up`之后配置的，而你是在之前（应该也不会有影响）。
```shell
ip link add veth0 type veth peer name veth1
ip link add veth2 type veth peer name veth3

**Comment by Plus-Jia** [comment]
This is an automatic reply, confirming that your e-mail was received.Thank you ! !Plus_Jia

**Comment by billnote** [comment]
```
# 配置命名空间1
$ ip netns exec ns1 ip link set veth1 up
$ ip netns exec ns1 ip addr add 172.16.0.1/24 dev veth1
```
这里是不是错了，激活/配置的应该veth0吧

然后应该还需要激活一下 veth1和veht3。 不然ping不通
```
ip link set veth1 up
ip link set veth3 up
```

**Comment by Plus-Jia** [comment]
This is an automatic reply, confirming that your e-mail was received.Thank you ! !Plus_Jia

**Comment by joy717** [comment]
现在的例子是有问题的.
只创建了一个veth pair，分别放到了ns1跟ns2，是没有办法在关联到br0上的.
正常的做法，应该是创建2个veth pair。 每一对的一端对应的ns1/ns2，领内外一端关联br0

有点奇怪，以前的例子应该是比较正确的？看评论区，应该也是之前的例子是正常的。

**Reply by isno** [reply]
Linux 各个发行版，或者内核版本，操作接口会有些不同， 我举的例子会有操作不通的情况。这本书例子不算多，能知道操作流程、设计思想就可以了

**Comment by Plus-Jia** [comment]
This is an automatic reply, confirming that your e-mail was received.Thank you ! !Plus_Jia
